File: fpga_configuration.py

# fpga_configuration.py

import sys
import datetime

sys.path.append('source/mcs_convert')
sys.path.append('source/erase')
sys.path.append('source/write_file')
sys.path.append('source/verify')
sys.path.append('source/reboot')

import mcs_convert
import erase_spi
import write_file_to_spi
import spi_verify
import reboot_start

def fpga_configuration() :
   
   config_file = open("config_folder/config_file.txt","r")
   
   mcsfile_list = list(config_file.readline())
   ip_addr_list = list(config_file.readline())
   udpport_list = list(config_file.readline())
   """
   print mcsfile_list
   print ip_addr_list
   print udpport_list
   """
   mcsfile_list.remove('\n')
   mcsfile_list.remove('\r')
   ip_addr_list.remove('\n')
   ip_addr_list.remove('\r')
   udpport_list.remove('\n')
   udpport_list.remove('\r')
   
   mcsfile = "".join(mcsfile_list)
   ip_addr = "".join(ip_addr_list)
   udpport = "".join(udpport_list)
   
   int_udpport = int(udpport)
   
   print mcsfile
   print ip_addr
   print int_udpport
   
   config_file.close()
   
   print "config start"

   start_d   = datetime.datetime.today()
   
   
   convert_judge = raw_input('Do you pass convert? (No/Yes) : ')
   
   if (convert_judge == 'No') :
      mcs_convert.mcs_convert(filename = mcsfile)
   elif (convert_judge == 'Yes') :
      print "pass convert"
   else :
      print "please input Yes or No"
   
   
   erase_spi.erase_spi(ip_addr = ip_addr, port = int_udpport)
   
   write_file_size = write_file_to_spi.write_file_to_spi(ip_addr = ip_addr, port = int_udpport, filename = mcsfile)
   
   spi_verify.spi_verify(ip_addr = ip_addr, port = int_udpport, file_size = write_file_size ,conv_file_name = mcsfile)
   
   
   reboot_judge = raw_input('Do you reboot FPGA? (Yes/No) : ')
   
   if (reboot_judge == 'Yes') :
      reboot_start.reboot_start(ip_addr = ip_addr, port = int_udpport)
   elif (reboot_judge == 'No') :
      print "Writing filedata to SPI is finished."
   else :
      print "please input Yes or No"
   
   print "Writing filedata to SPI finished."
   print "Please wait until FPGA reboot finished."

   end_d = datetime.datetime.today()
   print start_d
   print end_d



if __name__ == '__main__' :
   fpga_configuration()

File: fpga_configuration.cc
// fpga_configuration.
#include <iostream>
#include <string>
#include <sstream>
#include <ctime>
#include <vector>
#include "mcs_convert.h"
#include "erase_spi.h"
#include "write_file_to_spi.h"
#include "spi_verify.h"
#include "reboot_start.h"

using namespace std;
template <typename T>
vector<T> getlist(ifstream&);

int main()
{
  ifstream fin("config_folder/config_file.txt");
  if(!fin){
    cerr << "Cannot find configfile" << endl;
    return -1;
  }
  vector<string>  mcsfile_list=getlist<string>(fin);
  vector<string>  ip_addr_list=getlist<string>(fin);
  vector<unsigned int>     udpport_list=getlist<unsigned int>(fin);

  // for(int i=0;i<mcsfile_list.size();i++) cout << mcsfile_list[i] << " ";
  // cout << endl;
  // for(int i=0;i<ip_addr_list.size();i++) cout << ip_addr_list[i] << " ";
  // cout << endl;
  // for(int i=0;i<udpport_list.size();i++) cout << udpport_list[i] << " ";
  // cout << endl;
   
  unsigned int udpport = udpport_list[0];
   
  string mcsfile(mcsfile_list[0]); cout << mcsfile << endl;
  string ip_addr(ip_addr_list[0]); cout << ip_addr << endl;
  cout << udpport << endl;
   
  fin.close();

  cout << "config start" << endl;

  time_t start_d = time(0);

  string convert_judge;
  cout << "Do you pass convert? (No/Yes) : " << flush;
 ERROR_1:
  cin >> convert_judge;

  if (convert_judge == "No") 
    mcs_convert(mcsfile.c_str());
  else if (convert_judge == "Yes") 
    cout << "pass convert" << endl;
  else {
    cout << "please input Yes or No : " << flush;
    goto ERROR_1;
  }
  cout << "Erase start" << endl;
  erase_spi(ip_addr.c_str(), udpport);

  cout << "Write Start" <<endl;
  int write_file_size = write_file_to_spi(ip_addr.c_str(), udpport, mcsfile.c_str());
  cout << "Write Finished" <<endl;

  cout << "Verify Start" << endl;
  spi_verify(ip_addr.c_str(), udpport, write_file_size, mcsfile.c_str());

  string reboot_judge;
  cout << "Do you reboot FPGA? (Yes/No) : " << flush;
 ERROR_2:
  cin >> reboot_judge;
  
  if (reboot_judge == "Yes")
    reboot_start(ip_addr.c_str(), udpport);
  else if (reboot_judge == "No")
    cout << "Writing filedata to SPI is finished." << endl;
  else {
    cout << "please input Yes or No : " << flush;
    goto ERROR_2;
  }

  cout << "Writing filedata to SPI finished." << endl;
  cout << "Please wait until FPGA reboot finished." << endl;

  time_t end_d = time(0);
  cout << "start: " << ctime(&start_d) << flush;
  cout << "end:   " << ctime(&end_d) << flush;

}

template <typename T>
vector<T> getlist(ifstream &fin)
{
  vector<T> v;
  string buf;
  T buf2;
  getline(fin,buf);
  stringstream ss(buf);
  while(ss>>buf2)
    v.push_back(buf2);

  return v;
}
